src/groups/group.jl

```diff
@@ -0,0 +1,514 @@
+    inv!(::AbstractGroupManifold{Op}, y, x)
+    inv(::AbstractGroupManifold{Op}, x)
+    compose_left(::AbstractGroupManifold{Op}, x, y)
+    compose_left!(::AbstractGroupManifold{Op}, z, x, y)
+
+Note that a manifold is connected with an operation by wrapping it with
+a decorator, [`AbstractGroupManifold`](@ref). In typical cases the concrete
+wrapper [`GroupManifold`](@ref) can be used.
+"""
+abstract type AbstractGroupOperation end
+
+@doc doc"""
+    AbstractGroupManifold{<:AbstractGroupOperation} <: Manifold
+
+Abstract type for a Lie group, a group that is also a smooth manifold with a
+[smooth binary operation](@ref AbstractGroupOperation).
+`AbstractGroupManifold`s must implement at least [`inv`](@ref),
+[`identity`](@ref), [`compose_left`](@ref), and [`translate_diff`](@ref).
+Group manifolds by default forward metric-related operations to the wrapped
+manifold.
+"""
+abstract type AbstractGroupManifold{O<:AbstractGroupOperation} <: Manifold end
+
+"""
+    GroupManifold{M<:Manifold,O<:AbstractGroupOperation} <: AbstractGroupManifold{O}
+
+Decorator for a smooth manifold that equips the manifold with a group
+operation, thus making it a Lie group. See [`AbstractGroupManifold`](@ref) for
+more details.
+
+# Constructor
+
+    GroupManifold(manifold, op)
+"""
+struct GroupManifold{M<:Manifold,O<:AbstractGroupOperation} <: AbstractGroupManifold{O}
+    manifold::M
+    op::O
+end
+
+is_decorator_manifold(::GroupManifold) = Val(true)
+
+"""
+    ActionDirection
+
+Direction of action on a manifold, either [`LeftAction`](@ref) or
+[`RightAction`](@ref).
+"""
+abstract type ActionDirection end
+
+"""
+    LeftAction()
+
+Left action of a group on a manifold.
+"""
+struct LeftAction <: ActionDirection end
+
+"""
+    RightAction()
+
+Right action of a group on a manifold.
+"""
+struct RightAction <: ActionDirection end
+
+
+@doc doc"""
+    Identity(group::AbstractGroupManifold)
+
+The identity element of the group `group`.
+"""
+struct Identity{G<:AbstractGroupManifold}
+    group::G
+end
+
+(e::Identity)(x) = identity(e.group, x)
+
+@doc doc"""
+    inv(G::AbstractGroupManifold, x)
+
+Inverse $x^{-1}$ of an element $x$, such that
+$x \cdot x^{-1} = x^{-1} \cdot x = e$.
+"""
+function inv!(G::AbstractGroupManifold, y, x)
+    error("inv not implemented on $(typeof(G)) for point $(typeof(x))")
+end
+
+function inv!(G::AbstractGroupManifold, y, e::Identity)
+    indentity!(G, y)
+    return y
+end
+
+function inv(G::AbstractGroupManifold, x)
+    y = similar_result(G, inv, x)
+    inv!(G, y, x)
+    return y
+end
+
+inv(::AbstractGroupManifold, e::Identity) = e
+
+function identity!(G::AbstractGroupManifold, y, x)
+    error("identity! not implemented on $(typeof(G)) for points $(typeof(y)) and $(typeof(x))")
+end
+
+@doc doc"""
+    identity(G::AbstractGroupManifold, x)
+
+Identity element $e$, such that for any element $x$,
+$x \cdot e = e \cdot x = x$. The returned element is of a similar type to `x`.
```

Comment:
Use the same notation for group operation here as in the docstring for `AbstractGroupOperation`

Reply:
I think it's OK now.